src/types/mod.rs

```diff
@@ -99,8 +99,13 @@ macro_rules! pyobject_native_type_base(
             fn fmt(&self, f: &mut ::std::fmt::Formatter<'_>)
                    -> ::std::result::Result<(), ::std::fmt::Error>
             {
-                let s = self.str().or(::std::result::Result::Err(::std::fmt::Error))?;
-                f.write_str(&s.to_string_lossy())
+                if let ::std::result::Result::Ok(s) = self.str() {
+                    f.write_str(&s.to_string_lossy())
+                } else if let ::std::result::Result::Ok(name) = self.get_type().name() {
+                    ::std::write!(f, "<unprintable {} object>", name)
```

Comment:
If `self.str()` fails, it may be nice to try to include the error text in this message so that it helps users understand why the object was unprintable

Not sure exactly what that would look like, something like this maybe?

```
<repr for object of type {type} failed with: {exc_type} {exc_msg}>
```

Would be quite a lengthy repr though...

Reply:
is there any chance that could be a security risk - e.g. leaking information about the failure in a place where you wouldn't expect it?

Reply:
I fear that if one relies on `repr` to produce externally visible log messages, that concern already applies? Otherwise, if I know the possible types and that their `repr` is safe, I should also know that these errors do not happen, don't I?

Reply:
Well, this is using `str` not `repr`, which might generally be considered "more user facing".

Reply:
I don't think the distinction matters too much when sensitive information is involved, e.g. an API key stored as a `str` property of a Python class might be part of both its `str` and its `repr`?

Reply:
I agree with @adamreichold here that if there is sensitive information then it's the type's responsibility to protect that adequately, not something which this framework can do. I wouldn't see this as a security risk.

Besides, I've had an even better idea: this should probably report the error using `PyErr::write_unraisable`.

Reply:
@samuelcolvin Would you be up to bending this PR towards `PyErr::write_unraisable`?

Reply:
I'm a bit flat out right now, won't be able to look for a week or two.

I'm also a bit unclear what you need?

Reply:
Ok, I took the liberty of rebasing this and making use of `PyErr::write_unraisable` to report the error. I think this is ready to merge now and hence also removed the draft status.